Fixes stuff inside bolt12 decode

Harshit933 opened the pull request:
Decoding bolt12 now:

```
[harshit:~] lampo-cli --network regtest decode --invoice_str lno1qgsyxjtl6luzd9t3pr62xr7eemp6awnejusgf6gw45q75vcfqqqqqqqsespexwyy4tcadvgg89l9aljus6709kx235hhqrk6n8dey98uyuftzdqzrtkahuum7m56dxlnx8r6tffy54004l7kvs7pylmxx7xs4n54986qyqeeuhhunayntt50snmdkq4t7fzsgghpl69v9csgparek8kv7dlp5uqr8ymp5s4z9upmwr2s8xu020d45t5phqc8nljrq8gzsjmurzevawjz6j6rc95xwfvnhgfx6v4c3jha7jwynecrz3y092nn25ek4yl7xp9yu9ry9zqagt0ktn4wwvqg52v9ss9ls22sqyqqestzp2l6decpn87pq96udsvx
{
  "issuer": "No issuer id provided",
  "offer_id": "[52, 70, 8, 105, 84, 158, 55, 116, 140, 234, 171, 220, 255, 98, 132, 169, 130, 102, 193, 128, 82, 171, 42, 126, 158, 181, 161, 175, 10, 94, 91, 125]",
  "offer_chains": [
    "43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000"
  ],
  "description": "Description not provided",
  "offer_paths": [
    {
      "blinded_hops": [
        "0339e5efc9f4935ae8f84f6db02abf2450422e1fe8ac2e2080f479b1eccf37e1a7",
        "031448f2aa7355336a93fe304a4e14642881d42df65ceae73008a2985840bf8295"
      ],
      "blinding_points": "021aeddbf39bf6e9a69bf331c7a5a524a55efaffd6643c127f66378d0ace9529f4"
    }
  ],
  "network": "testnet"
}
```

If the bolt11/bolt12 is invalid:
```
[harshit:~] lampo-cli --network regtest decode --invoice_str lno1qgsyxjtl6luzd9t3pr62xr7eemp6awnejusgf6gw45q75vcfqqqqqqqsespexwyy4tcadvgg89l9aljus6709kx235hhqrk6n8dey98uyuftzdqzrtkahuum7m56dxlnx8r6tffy54004l7kvs7pylmxx7xs4n54986qyqeeuhhunayntt50snmdkq4t7fzsgghpl69v9csgparek8kv7dlp5uqr8ymp5s4z9upmwr2s8xu020d45t5phqc8nljrq8gzsjmurzevawjz6j6rc95xwfvnhgfx6v4c3jha7jwynecrz3y092nn25ek4yl7xp9yu9ry9zqagt0ktn4wwvqg52v9ss9ls22sqyqqestzp2l6decpn87pq96uds
{
  "code": -1,
  "message": "Not able to decode invoice",
  "data": null
}
```

UPDATE:

Running `lampo-cli --network regtest decode --invoice_str lno1qgsyxjtl6luzd9t3pr62xr7eemp6awnejusgf6gw45q75vcfqqqqqqqsespexwyy4tcadvgg89l9aljus6709kx235hhqrk6n8dey98uyuftzdqzrtkahuum7m56dxlnx8r6tffy54004l7kvs7pylmxx7xs4n54986qyqeeuhhunayntt50snmdkq4t7fzsgghpl69v9csgparek8kv7dlp5uqr8ymp5s4z9upmwr2s8xu020d45t5phqc8nljrq8gzsjmurzevawjz6j6rc95xwfvnhgfx6v4c3jha7jwynecrz3y092nn25ek4yl7xp9yu9ry9zqagt0ktn4wwvqg52v9ss9ls22sqyqqestzp2l6decpn87pq96udsvx`

Gives the following output:
```
{
  "issuer_id": null,
  "offer_id": "34460869549e37748ceaabdcff6284a98266c18052ab2a7e9eb5a1af0a5e5b7d",
  "offer_chains": [
    "43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000"
  ],
  "description": null,
  "offer_paths": [
    {
      "blinded_hops": [
        "0339e5efc9f4935ae8f84f6db02abf2450422e1fe8ac2e2080f479b1eccf37e1a7",
        "031448f2aa7355336a93fe304a4e14642881d42df65ceae73008a2985840bf8295"
      ],
      "blinding_points": "021aeddbf39bf6e9a69bf331c7a5a524a55efaffd6643c127f66378d0ace9529f4"
    }
  ],
  "network": "testnet"
}
```
See commits for more info.
Fixes #270 #318

vincenzopalazzo commented on lampod/src/jsonrpc/offchain.rs:
```diff
@@ -80,40 +79,69 @@ pub fn json_decode_invoice(ctx: &LampoDaemon, request: &json::Value) -> Result<j
+            .map(|issuer| issuer.to_string())
+            .unwrap_or_else(|| "No issuer id provided".to_string());
+
+        let bolt12_invoice = Bolt12InvoiceInfo {
+            offer_id,
+            network: network.unwrap().to_string(),
+            description: desc,
+            offer_chains: chains,
+            offer_paths: paths,
+            issuer: issuer_id,
+        };
```
Do you think that it is possible to implement the `From<ldk::Bolt12Offer> for Bolt12InvoiceInfo`? IDK what is the ldk type name bit should be good to have something like `let bolt12invoice = offer.info()`

Harshit933 replied:
Done inside d7ceae28b8acaf599a2e158ca7317844357e295b

vincenzopalazzo commented on lampod/src/jsonrpc/offchain.rs:
```diff
@@ -80,40 +79,69 @@ pub fn json_decode_invoice(ctx: &LampoDaemon, request: &json::Value) -> Result<j
+
+        //FIXME: Convert this to hex
+        let offer_id = format!("{:?}", offer.id().0);
+        let desc = offer
+            .description()
+            .map(|desc| desc.to_string())
+            .unwrap_or_else(|| "Description not provided".to_string());
+
+        let issuer_id = offer
+            .issuer()
+            .map(|issuer| issuer.to_string())
+            .unwrap_or_else(|| "No issuer id provided".to_string());
```
When a type is None, do not force to be not none. Keep it `Option` otherwise the user of our API need to check the ID.

If there is no description, leave it None.

Harshit933 replied:
Done d7ceae28b8acaf599a2e158ca7317844357e295b

vincenzopalazzo commented:
Plus `offer_id` is a vector of bytes, please convert it to `hex::(offer_id)` there is no need to print a vector of byte to the user

vincenzopalazzo commented:
@Harshit933 can you post another time the `lampo-cli --network regtest decode --invoice_str lno1qgsyxjtl6luzd9t3pr62xr7eemp6awnejusgf6gw45q75vcfqqqqqqqsespexwyy4tcadvgg89l9aljus6709kx235hhqrk6n8dey98uyuftzdqzrtkahuum7m56dxlnx8r6tffy54004l7kvs7pylmxx7xs4n54986qyqeeuhhunayntt50snmdkq4t7fzsgghpl69v9csgparek8kv7dlp5uqr8ymp5s4z9upmwr2s8xu020d45t5phqc8nljrq8gzsjmurzevawjz6j6rc95xwfvnhgfx6v4c3jha7jwynecrz3y092nn25ek4yl7xp9yu9ry9zqagt0ktn4wwvqg52v9ss9ls22sqyqqestzp2l6decpn87pq96udsvx`

And/or probably write a test to make sure that what we should have as a hex string it is not a vector of bytes?

otherwise LGMT Thanks :)

Harshit933 commented:
> @Harshit933 can you post another time the `lampo-cli --network regtest decode --invoice_str lno1qgsyxjtl6luzd9t3pr62xr7eemp6awnejusgf6gw45q75vcfqqqqqqqsespexwyy4tcadvgg89l9aljus6709kx235hhqrk6n8dey98uyuftzdqzrtkahuum7m56dxlnx8r6tffy54004l7kvs7pylmxx7xs4n54986qyqeeuhhunayntt50snmdkq4t7fzsgghpl69v9csgparek8kv7dlp5uqr8ymp5s4z9upmwr2s8xu020d45t5phqc8nljrq8gzsjmurzevawjz6j6rc95xwfvnhgfx6v4c3jha7jwynecrz3y092nn25ek4yl7xp9yu9ry9zqagt0ktn4wwvqg52v9ss9ls22sqyqqestzp2l6decpn87pq96udsvx`

Updated the description also added a test in 593fe3ef0188caf7ab2befb791b89d6269025c49

vincenzopalazzo commented on lampod/src/jsonrpc/offchain.rs:
```diff
@@ -80,40 +79,69 @@ pub fn json_decode_invoice(ctx: &LampoDaemon, request: &json::Value) -> Result<j
+            .collect::<Vec<BlindedPath>>();
+
+        //FIXME: Convert this to hex
+        let offer_id = format!("{:?}", offer.id().0);
```
Ok this is still not correct!

Something like ` let offer_id = hex::encode(offer.id().0);` should work.

If you change this part of the code later, you should rework the commit history

Harshit933 replied:
This was done previously, but I cleaned up the git history now.

vincenzopalazzo commented on lampod/src/jsonrpc/offchain.rs:
```diff
@@ -80,40 +79,69 @@ pub fn json_decode_invoice(ctx: &LampoDaemon, request: &json::Value) -> Result<j
+        let desc = offer
+            .description()
+            .map(|desc| desc.to_string())
+            .unwrap_or_else(|| "Description not provided".to_string());
```
The description needs to be None if it is not provided. what happens if someone put the string `Description not provide` inside the description?

Please use Rust as a properly typed language

Harshit933 replied:
This was also fixed inside 68d57c905baedf14224e5bcafed8cd088fcb6d06

vincenzopalazzo commented on lampod/src/jsonrpc/offchain.rs:
```diff
@@ -80,40 +79,69 @@ pub fn json_decode_invoice(ctx: &LampoDaemon, request: &json::Value) -> Result<j
+        let issuer_id = offer
+            .issuer()
+            .map(|issuer| issuer.to_string())
+            .unwrap_or_else(|| "No issuer id provided".to_string());
```
same here, so this commit should be reworked if you change it later

vincenzopalazzo commented on lampo-common/src/model/invoice.rs:
```diff
@@ -72,14 +73,57 @@ pub mod response {
+            let network = offer
+                .chains()
+                .first()
+                .map(|hash| Network::from_chain_hash(*hash))
+                .unwrap_or(Some(Network::Bitcoin));
```
Add a reference to the bolts where it is telling that if the chain is none the network is implied to bitcoin

Harshit933 replied:
Done inside 68d57c905baedf14224e5bcafed8cd088fcb6d06

vincenzopalazzo commented on lampo-common/src/model/invoice.rs:
```diff
@@ -72,14 +73,57 @@ pub mod response {
         pub network: String,
     }
 
+    impl From<LDKOffer> for Bolt12InvoiceInfo {
```
Nice trick!

vincenzopalazzo reviewed (changes requested):
Ok I did another  review, thanks for doing it.

I think the git history need some kind of rework (squash the commit with the one that are no longer true) and then we are ready to go :)

Thanks for doing it

Harshit933 commented on lampo-testing/src/lib.rs:
```diff
@@ -116,9 +116,7 @@ impl LampoTesting {
-        server
-            .add_rpc("decode_invoice", json_decode_invoice)
-            .unwrap();
+        server.add_rpc("decode", json_decode_invoice).unwrap();
```
Done inside ea1fcf31367af1008ea14d8a58c687efb35a5140

Harshit933 commented on tests/tests/src/lampo_tests.rs:
```diff
@@ -505,10 +505,28 @@ pub fn decode_offer() -> error::Result<()> {
+    )?;
+
+    // Checks if all the character inside the offer_id is hex.
+    let res = decode.offer_id.chars().all(|c| c.is_ascii_hexdigit());
+    assert!(res);
```
Done inside 6a02c694bbab9c52dcc5e3e451c01e523a023c0d

vincenzopalazzo commented on lampo-common/src/model/invoice.rs:
```diff
@@ -70,6 +71,66 @@ pub mod response {
+                .map(|chain| chain.to_string())
+                .collect::<Vec<String>>();
+
+            // Reference: https://github.com/lightning/bolts/blob/master/12-offer-encoding.md
```
```rust
// Reference: https://github.com/lightning/bolts/blob/master/12-offer-encoding.md#requirements-for-offers
// if the chain for the invoice is not solely bitcoin:
// MUST specify offer_chains the offer is valid for.
// otherwise:
// SHOULD omit offer_chains, implying that bitcoin is only chain.
```

Better quote the spec in this way you know if this code is outdated when and if some part of the spec change.

Harshit933 replied:
Done

vincenzopalazzo commented on lampo-common/src/model/invoice.rs:
```diff
@@ -70,6 +71,66 @@ pub mod response {
+            let network = offer
+                .chains()
+                .first()
+                .map(|hash| Network::from_chain_hash(*hash))
+                .unwrap_or(Some(Network::Bitcoin));
```
we can find a way to return the following `.unwrap_or(Network::Bitcoin);`

In this way we avoid the unwrap down in the code!

Harshit933 replied:
Fixed

vincenzopalazzo commented on lampod/src/jsonrpc/offchain.rs:
```diff
@@ -80,40 +77,24 @@ pub fn json_decode_invoice(ctx: &LampoDaemon, request: &json::Value) -> Result<j
-    };
-
-    Ok(json::to_value(&invoice)?)
+        Err(Error::Rpc(RpcError {
+            code: -1,
+            message: "Not able to decode invoice".to_string(),
+            data: None,
+        }))
+    }
```
Do we have an `error!` macros to use?

Harshit933 replied:
Fixed

vincenzopalazzo commented on lampo-testing/src/lib.rs:
```diff
@@ -116,9 +116,7 @@ impl LampoTesting {
-        server
-            .add_rpc("decode_invoice", json_decode_invoice)
-            .unwrap();
+        server.add_rpc("decode", json_decode_invoice).unwrap();
```
wondering if we need to change this also in the `lampod-cli`

vincenzopalazzo reviewed (commented):
Concept ACK 6a02c694bbab9c52dcc5e3e451c01e523a023c0d

Small comments on the code still, but we are good to go! thanks for working on it

vincenzopalazzo commented on lampo-common/src/model/invoice.rs:
```diff
@@ -89,12 +89,16 @@ pub mod response {
-                .map(|hash| Network::from_chain_hash(*hash))
-                .unwrap_or(Some(Network::Bitcoin));
+                .and_then(|hash| Network::from_chain_hash(*hash))
+                .or(Some(Network::Bitcoin));
```
Good also for a folloup PR, but I would like to leave this Optional and not try to force the type here!

If there is any `chain`, will remain `None`

vincenzopalazzo commented on tests/tests/src/lampo_tests.rs:
```diff
@@ -509,3 +509,28 @@ pub fn decode_offer() -> error::Result<()> {
+        decode.offer_id,
+        "34460869549e37748ceaabdcff6284a98266c18052ab2a7e9eb5a1af0a5e5b7d"
+    );
+    // Checks if all the character inside the offer_id is hex.
+    let res = decode.offer_id.chars().all(|c| c.is_ascii_hexdigit());
+    assert!(res);
```
we can remove this code, I do not understand it, so better remove it (also in a follup PR)

vincenzopalazzo reviewed (approved):
LGTM
